crates/miden-lib/asm/note_scripts/P2IDE.masm

```diff
@@ -0,0 +1,201 @@
+#! - Any of the following conditions is true:
+#!   - The account ID against which the transaction is executed matches the note's target account id.
+#!   - The account ID against which the transaction is executed matches the note's sender account id and
+#!     the transaction's reference block number is greater than or equal to the note's reclaim block height,
+#!     or the reclaim block height is 0, which means it is disabled.
+#!
+#! Requires that the account exposes:
+#! - miden::contracts::wallets::basic::receive_asset procedure.
+#!
+#! Inputs:  []
+#! Outputs: []
+#!
+#! Note inputs are assumed to be as follows:
+#! - target_account_id is the ID of the account for which the note is intended.
+#! - reclaim_block_height is the block height at which the note can be reclaimed by the sender.
+#! - timelock_block_height is the block height at which the note can be consumed by the target.
+#!
+#! Panics if:
+#! - The account does not expose miden::contracts::wallets::basic::receive_asset procedure.
+#! - The note is consumed before the timelock expired, i.e. the transaction's reference block
+#!   number is less than the timelock block height.
+#! - Before reclaim block height: the account ID of the executing account is not equal to the specified
+#!   account ID.
+#! - At and after reclaim block height: the account ID of the executing account is not equal to
+#!   the specified account ID or sender account ID.
+#! - The same non-fungible asset already exists in the account.
+#! - Adding a fungible asset would result in an amount overflow, i.e., the total amount would be
+#!   greater than 2^63.
+begin
+    # store the note inputs to memory starting at address 0
+    push.0 exec.note::get_inputs
+    # => [num_inputs, inputs_ptr]
+
+    # make sure the number of inputs is 4
+    eq.4 assert.err=ERR_P2IDE_WRONG_NUMBER_OF_INPUTS
+    # => [inputs_ptr]
+
+    # read the reclaim block height, timelock_block_height, and target account ID from the note inputs
+    mem_loadw
+    # => [timelock_block_height, reclaim_block_height, target_account_id_prefix, target_account_id_suffix]
+
+    # fails if note is locked
+    exec.verify_unlocked
+    # => [reclaim_block_height, target_account_id_prefix, target_account_id_suffix]
+
+    # get current account id
+    exec.account::get_id dup.1 dup.1
+    # => [account_id_prefix, account_id_suffix, account_id_prefix, account_id_suffix, reclaim_block_height, target_account_id_prefix, target_account_id_suffix]
+
+    # determine if the current account is the target account
+    movup.6 movup.6 exec.account::is_id_equal
+    # => [is_target, account_id_prefix, account_id_suffix, reclaim_block_height]
+
+    if.true
+        # we can safely consume the note since the current account is the target of the note
+        dropw exec.add_note_assets_to_account
+        # => []
+
+    else
+        # checks if current account is sender and if reclaim is enabled
+        exec.consume_by_sender
```

Comment:
Nit: I think it might be more apt to call this `reclaim_note` because at this point we have not yet established that the account that is trying to consume the note is in fact the sender. Instead the procedure asserts that. `consume_by_sender` suggests to me that we already know the consuming account is the sender.